### PR TITLE
Consumer block until registered

### LINKED ISSUE #1392: Consumer needs to check registration status after calling register on transactor
The consumer will call transactors register endpoint. The transactor will send the TX to BC but will not ensure that it is completed successfully. It is up to the consumer to watch for registration status changes and decide if transactor needs to be called again.

Big question here though is if this should be done via a goroutine on node side, or is it up to the consumer(cli, web) to check if the identity is registered or not. The status endpoint already returns the identity registration status, so it can be used as an indicator here.

### PR DESCRIPTION
**Pull request by vkuznecovas**
Registry is now fully responsible for the registration flow of an identity.

Consumer will now be unable to attempt a connection before registering.

Closes #1392

### CODE REVIEW

File: ci/packages/package.go [hunk 1]
```diff
@@ -222,7 +222,7 @@ func checkVend() error {
 		fmt.Println("Tool 'vend' already installed")
 		return nil
 	}
-	err := goGet("github.com/nomad-software/vend")
+	err := goGet("github.com/mysteriumnetwork/vend")
```
Comment (Waldz) [inline ci/packages/package.go]: Guys, why these changes are generated here?

File: ci/packages/package.go [hunk 2]
```diff
@@ -213,7 +213,7 @@ func PackageDockerSwaggerRedoc() error {
 // does not support go modules yet and go mod vendor does not include c dependencies.
 func vendordModules() error {
 	mg.Deps(checkVend)
-	return sh.RunV("vend")
+	return sh.RunV("vend", "-replace", "golang.zx2c4.com/wireguard=github.com/mysteriumnetwork/wireguard-go")
```
Comment (Waldz) [inline ci/packages/package.go]: and here

File: core/discovery/discovery.go [hunk 3]
```diff
@@ -170,26 +170,29 @@ func (d *Discovery) stopLoop() {
 	d.mu.RUnlock()
 }
 
+func (d *Discovery) handleRegistrationEvent(rep registry.RegistrationEventPayload) {
+	log.Debug().Msgf("registration event received for %v", rep.ID.Address)
```
Comment (soffokl) [inline core/discovery/discovery.go]:
```suggestion
	log.Debug().Msgf("Registration event received for %v", rep.ID.Address)
```

File: core/discovery/discovery.go [hunk 4]
```diff
@@ -170,26 +170,29 @@ func (d *Discovery) stopLoop() {
...
+	if rep.ID.Address != d.ownIdentity.Address {
+		log.Debug().Msgf("identity missmatch for registration. Expected %v got %v", d.ownIdentity.Address, rep.ID.Address)
```
Comment (soffokl) [inline core/discovery/discovery.go]:
```suggestion
		log.Debug().Msgf("Identity missmatch for registration. Expected %v got %v", d.ownIdentity.Address, rep.ID.Address)
```

File: core/discovery/discovery.go [hunk 5]
```diff
@@ -170,26 +170,29 @@ func (d *Discovery) stopLoop() {
...
+
+	switch rep.Status {
+	case registry.RegisteredProvider:
+		log.Info().Msg("Identity registered, proceeding with proposal registration")
+		d.changeStatus(RegisterProposal)
+	case registry.RegistrationError:
+		log.Info().Msg("Cancelled identity registration")
+		d.changeStatus(IdentityRegisterFailed)
+	default:
+		log.Info().Msgf("received status %v ignoring", rep.Status)
```
Comment (soffokl) [inline core/discovery/discovery.go]:
```suggestion
		log.Info().Msgf("Received status %v ignoring", rep.Status)
```

File: core/discovery/discovery.go [hunk 6]
```diff
@@ -226,20 +229,20 @@ func (d *Discovery) unregisterProposal() {
...
 		return
 	}
-
-	if !registered {
-		// TODO: Maybe register here?
+	switch status {
+	case identity_registry.RegisteredProvider:
+		d.changeStatus(RegisterProposal)
+	default:
 		log.Info().Msgf("identity %s not registered, delaying proposal registration until identity is registered", d.ownIdentity.Address)
```
Comment (soffokl) [inline core/discovery/discovery.go]:
```suggestion
		log.Info().Msgf("Identity %s not registered, delaying proposal registration until identity is registered", d.ownIdentity.Address)
```

File: identity/registry/provider_registrar.go [hunk 7]
```diff
@@ -146,20 +146,20 @@ func (pr *ProviderRegistrar) delayedRequeue(qe queuedEvent) {
...
 	}
 
-	if registered {
+	switch registered {
+	case RegisteredProvider:
 		log.Info().Msgf("provider %q already registered on bc, skipping", qe.event.ProviderID)
 		pr.registeredIdentities[qe.event.ProviderID] = struct{}{}
 		return nil
+	default:
+		log.Info().Msgf("provider %q not registered on BC, will register", qe.event.ProviderID)
```
Comment (soffokl) [inline identity/registry/provider_registrar.go]:
```suggestion
		log.Info().Msgf("Provider %q not registered on BC, will register", qe.event.ProviderID)
```

File: identity/registry/provider_registrar.go [hunk 8]
```diff
@@ -146,20 +146,20 @@ func (pr *ProviderRegistrar) delayedRequeue(qe queuedEvent) {
...
-	registered, err := pr.bc.IsRegisteredAsProvider(pr.cfg.AccountantAddress, pr.cfg.RegistryAddress, common.HexToAddress(qe.event.ProviderID))
+	registered, err := pr.registrationStatusChecker.GetRegistrationStatus(identity.FromAddress(qe.event.ProviderID))
 	if err != nil {
 		return errors.Wrap(err, "could not check registration status on BC")
 	}
 
-	if registered {
+	switch registered {
+	case RegisteredProvider:
 		log.Info().Msgf("provider %q already registered on bc, skipping", qe.event.ProviderID)
```
Comment (soffokl) [inline identity/registry/provider_registrar.go]:
```suggestion
		log.Info().Msgf("Provider %q already registered on bc, skipping", qe.event.ProviderID)
```

File: identity/registry/registry_contract.go [hunk 9]
```diff
@@ -19,17 +19,44 @@ package registry
...
+	once                       sync.Once
+	registrationCompletionLock sync.Mutex
+	publisher                  eventbus.Publisher
+	lock                       sync.Mutex
+}
+
 // NewIdentityRegistryContract creates identity registry service which uses blockchain for information
-func NewIdentityRegistryContract(contractBackend bind.ContractBackend, registryAddress, accountantAddress common.Address) (*contractRegistry, error) {
+func NewIdentityRegistryContract(contractBackend bind.ContractBackend, registryAddress, accountantAddress common.Address, registryStorage registryStorage, publisher eventbus.Publisher) (*contractRegistry, error) {
+	log.Info().Msgf("using registryAddress %v accountantAddress %v", registryAddress.Hex(), accountantAddress.Hex())
```
Comment (soffokl) [inline identity/registry/registry_contract.go]:
```suggestion
	log.Info().Msgf("Using registryAddress %v accountantAddress %v", registryAddress.Hex(), accountantAddress.Hex())
```

File: identity/registry/registry_contract.go [hunk 10]
```diff
@@ -42,57 +69,125 @@ func NewIdentityRegistryContract(contractBackend bind.ContractBackend, registryA
...
+		Identity:           id,
+		RegistrationStatus: statusBC,
+	})
+	return statusBC, errors.Wrap(err, "could not store registration status")
 }
 
-// RegistrationEvent describes registration events
-type RegistrationEvent int
+func (registry *contractRegistry) handleNodeEvent(ev event.Payload) {
+	log.Debug().Msgf("event received %v", ev)
```
Comment (soffokl) [inline identity/registry/registry_contract.go]:
```suggestion
	log.Debug().Msgf("Event received %v", ev)
```

File: identity/registry/registry_contract.go [hunk 11]
```diff
@@ -42,57 +69,125 @@ func NewIdentityRegistryContract(contractBackend bind.ContractBackend, registryA
...
-	Registered RegistrationEvent = 0
-	Cancelled  RegistrationEvent = 1
-)
+func (registry *contractRegistry) handleRegistrationEvent(ev IdentityRegistrationRequest) {
+	registry.lock.Lock()
+	defer registry.lock.Unlock()
+
+	status, err := registry.storage.Get(identity.FromAddress(ev.Identity))
+	if err != nil && err != ErrNotFound {
+		log.Error().Err(err).Msg("could not get status from local db")
```
Comment (soffokl) [inline identity/registry/registry_contract.go]:
```suggestion
		log.Error().Err(err).Msg("Could not get status from local db")
```

File: identity/registry/registry_contract.go [hunk 12]
```diff
@@ -42,57 +69,125 @@ func NewIdentityRegistryContract(contractBackend bind.ContractBackend, registryA
...
+			return
+		}
+	}
 
-	stopLoop := make(chan bool)
-	unsubscribe = func() {
-		// cancel (stop) identity registration loop
-		stopLoop <- true
+	if status.RegistrationStatus == RegisteredProvider {
+		log.Info().Msgf("identity %q already fully registered, skipping", ev.Identity)
```
Comment (soffokl) [inline identity/registry/registry_contract.go]:
```suggestion
		log.Info().Msgf("Identity %q already fully registered, skipping", ev.Identity)
```

File: identity/registry/registry_contract.go [hunk 13]
```diff
@@ -104,22 +199,176 @@ func (registry *contractRegistry) SubscribeToRegistrationEvent(id identity.Ident
 		sink := make(chan *bindings.RegistryRegisteredIdentity)
 		subscription, err := registry.filterer.WatchRegisteredIdentity(filterOps, sink, userIdentities, accountantIdentities)
 		if err != nil {
-			registrationEvent <- Cancelled
-			log.Error().Err(err).Msg("")
+			registry.publisher.Publish(RegistrationEventTopic, RegistrationEventPayload{
+				ID:     identity,
+				Status: RegistrationError,
+			})
+			log.Error().Err(err).Msg("could not register to identity events")
```
Comment (soffokl) [inline identity/registry/registry_contract.go]:
```suggestion
			log.Error().Err(err).Msg("Could not register to identity events")
```

File: identity/registry/registry_contract.go [hunk 14]
```diff
@@ -104,22 +199,176 @@ func (registry *contractRegistry) SubscribeToRegistrationEvent(id identity.Ident
...
+
+			updateErr := registry.storage.Store(StoredRegistrationStatus{
+				Identity:           identity,
+				RegistrationStatus: RegistrationError,
+			})
+			if updateErr != nil {
+				log.Error().Err(updateErr).Msg("could not store registration status")
```
Comment (soffokl) [inline identity/registry/registry_contract.go]:
```suggestion
				log.Error().Err(updateErr).Msg("Could not store registration status")
```

File: identity/registry/registry_contract.go [hunk 15]
```diff
@@ -104,22 +199,176 @@ func (registry *contractRegistry) SubscribeToRegistrationEvent(id identity.Ident
...
+
+func (registry *contractRegistry) handleStop() {
+	registry.once.Do(func() {
+		log.Info().Msg("Stopping registry...")
+		close(registry.stop)
+	})
+}
+
+func (registry *contractRegistry) handleStart() {
+	log.Info().Msg("starting registry...")
```
Comment (soffokl) [inline identity/registry/registry_contract.go]: I have stopped marking all logs to be capitalized in the PR.
I have created an issue that will help to keep it consistent in the future: https://github.com/mysteriumnetwork/node/issues/1414

File: ci/packages/package.go (hunk 1 above)
Comment (vkuznecovas) [reply ci/packages/package.go]: https://github.com/mysteriumnetwork/node/pull/1410
